### ISSUE TITLE
crash in dttree while loading autoload functions

### ISSUE THREAD

**Issue by sowmya573** [ksh93/ksh#323]
In particular cases, when autoload functions are getting loaded in my application, ksh93 crashes in dttree while trying to 
do a "dtmatch" on autoload function. Seems like its dumping core while accessing the invalid address in dttree(). Could be a corruption in fpathdict tree. Note, i observed that it doesnt happen with single autoload function, only when multiple autoload functions are invoked.

Here is the stack trace:

Segmentation fault in . at 0xde0c
0x000000000000de0c warning: Unable to access address 0xde0c from core
(dbx) t
.() at 0xde0c
dttree(dt = 0x0a00000000110798, obj = (nil), type = 512), line 462 in "dttree.c"
funload(shp = 0x0000000110018aa8, fno = 3, name = "KLIB_HACMP_is_known_node"), line 609 in "path.c"
path_search(shp = 0x0000000110018aa8, name = "KLIB_HACMP_is_known_node", oldpp = (nil), flag = 1), line 737 in "path.c"
unnamed block in sh_exec(t = 0x0a000000000e8960, flags = 516), line 1284 in "xec.c"
unnamed block in sh_exec(t = 0x0a000000000e8960, flags = 516), line 1284 in "xec.c"
unnamed block in sh_exec(t = 0x0a000000000e8960, flags = 516), line 1284 in "xec.c"
sh_exec(t = 0x0a000000000e8960, flags = 516), line 1284 in "xec.c"
unnamed block in sh_exec(t = 0x0a000000000e8ee0, flags = 516), line 2224 in "xec.c"
sh_exec(t = 0x0a000000000e8ee0, flags = 516), line 2224 in "xec.c"
unnamed block in sh_exec(t = 0x0a000000000e8840, flags = 4), line 2369 in "xec.c"
unnamed block in sh_exec(t = 0x0a000000000e8840, flags = 4), line 2369 in "xec.c"
sh_exec(t = 0x0a000000000e8840, flags = 4), line 2369 in "xec.c"
unnamed block in sh_exec(t = 0x0a000000000e9820, flags = 4), line 2224 in "xec.c"
sh_exec(t = 0x0a000000000e9820, flags = 4), line 2224 in "xec.c"
unnamed block in sh_exec(t = 0x0a000000000e8138, flags = 4), line 2543 in "xec.c"
sh_exec(t = 0x0a000000000e8138, flags = 4), line 2543 in "xec.c"
unnamed block in sh_exec(t = 0x0a00000000195720, flags = 6), line 2224 in "xec.c"
sh_exec(t = 0x0a00000000195720, flags = 6), line 2224 in "xec.c"
unnamed block in sh_funscope(argn = 8, argv = 0x0a00000000017198, fun = (nil), arg = 0x0fffffffffff7c80, execflg = 4), line 3455 in "xec.c"
sh_funscope(argn = 8, argv = 0x0a00000000017198, fun = (nil), arg = 0x0fffffffffff7c80, execflg = 4), line 3455 in "xec.c"
sh_funct(shp = 0x0000000110018aa8, np = 0x0a0000000010ffc0, argn = 8, argv = 0x0a00000000017198, envlist = 0x0a000000000689b0, execflg = 4), line 3552 in "xec.c"
unnamed block in sh_exec(t = 0x0a000000000689d0, flags = 4), line 1573 in "xec.c"
unnamed block in sh_exec(t = 0x0a000000000689d0, flags = 4), line 1573 in "xec.c"
unnamed block in sh_exec(t = 0x0a000000000689d0, flags = 4), line 1573 in "xec.c"
unnamed block in sh_exec(t = 0x0a000000000689d0, flags = 4), line 1573 in "xec.c"
sh_exec(t = 0x0a000000000689d0, flags = 4), line 1573 in "xec.c"
unnamed block in sh_exec(t = 0x0a00000000068d00, flags = 6), line 2224 in "xec.c"
sh_exec(t = 0x0a00000000068d00, flags = 6), line 2224 in "xec.c"
unnamed block in sh_funscope(argn = 4, argv = 0x0a00000000016b50, fun = (nil), arg = 0x0fffffffffff9690, execflg = 4), line 3455 in "xec.c"
sh_funscope(argn = 4, argv = 0x0a00000000016b50, fun = (nil), arg = 0x0fffffffffff9690, execflg = 4), line 3455 in "xec.c"
sh_funct(shp = 0x0000000110018aa8, np = 0x0a00000000067e50, argn = 4, argv = 0x0a00000000016b50, envlist = 0x0a00000000016a48, execflg = 4), line 3552 in "xec.c"
unnamed block in sh_exec(t = 0x0a00000000016a68, flags = 4), line 1573 in "xec.c"
unnamed block in sh_exec(t = 0x0a00000000016a68, flags = 4), line 1573 in "xec.c"
unnamed block in sh_exec(t = 0x0a00000000016a68, flags = 4), line 1573 in "xec.c"
unnamed block in sh_exec(t = 0x0a00000000016a68, flags = 4), line 1573 in "xec.c"
sh_exec(t = 0x0a00000000016a68, flags = 4), line 1573 in "xec.c"
sh_eval(iop = 0x0a00000000097e98, mode = 0), line 631 in "xec.c"
b_eval(argc = 6, argv = 0x0a00000000016448, context = 0x0000000110019190), line 216 in "misc.c"
unnamed block in sh_exec(t = 0x0a000000000cf508, flags = 4), line 1407 in "xec.c"
unnamed block in sh_exec(t = 0x0a000000000cf508, flags = 4), line 1407 in "xec.c"
unnamed block in sh_exec(t = 0x0a000000000cf508, flags = 4), line 1407 in "xec.c"
unnamed block in sh_exec(t = 0x0a000000000cf508, flags = 4), line 1407 in "xec.c"
sh_exec(t = 0x0a000000000cf508, flags = 4), line 1407 in "xec.c"
unnamed block in sh_exec(t = 0x0a000000000cfcd8, flags = 4), line 2224 in "xec.c"
sh_exec(t = 0x0a000000000cfcd8, flags = 4), line 2224 in "xec.c"
unnamed block in sh_exec(t = 0x0a000000000ca8a8, flags = 4), line 2543 in "xec.c"
sh_exec(t = 0x0a000000000ca8a8, flags = 4), line 2543 in "xec.c"
unnamed block in sh_exec(t = 0x0a000000000d00f8, flags = 6), line 2224 in "xec.c"
sh_exec(t = 0x0a000000000d00f8, flags = 6), line 2224 in "xec.c"
unnamed block in sh_funscope(argn = 5, argv = 0x0a00000000015d88, fun = (nil), arg = 0x0fffffffffffd2a0, execflg = 4), line 3455 in "xec.c"
sh_funscope(argn = 5, argv = 0x0a00000000015d88, fun = (nil), arg = 0x0fffffffffffd2a0, execflg = 4), line 3455 in "xec.c"
sh_funct(shp = 0x0000000110018aa8, np = 0x0a000000000e6d70, argn = 5, argv = 0x0a00000000015d88, envlist = 0x0a00000000015c48, execflg = 4), line 3552 in "xec.c"
unnamed block in sh_exec(t = 0x0a00000000015c68, flags = 4), line 1573 in "xec.c"
unnamed block in sh_exec(t = 0x0a00000000015c68, flags = 4), line 1573 in "xec.c"
unnamed block in sh_exec(t = 0x0a00000000015c68, flags = 4), line 1573 in "xec.c"
unnamed block in sh_exec(t = 0x0a00000000015c68, flags = 4), line 1573 in "xec.c"
sh_exec(t = 0x0a00000000015c68, flags = 4), line 1573 in "xec.c"
exfile(shp = 0x0000000110018aa8, iop = 0x0a0000000002ed30, fno = 10), line 600 in "main.c"
sh_main(ac = 4, av = 0x0fffffffffffe530, userinit = (nil)), line 373 in "main.c"
pmain.main(argc = 4, argv = 0x0fffffffffffe530), line 59 in "pmain.c"
(dbx)

**Comment by hyenias** [comment]
I am trying to reproduce the above mentioned bug and at present unable to do so. Can you please use/adjust the following scripts to create a reproducer for us?

```ksh
$ cat func1
function func1 {
  echo func1 has executed.
}
$ cat func2
function func2 {
  echo func2 has worked.
}
$ cat autofunc.ksh
echo Testing autoload functions
echo KSH VER: $KSH_VERSION
typeset -fu func1
typeset -fu func2
typeset -f func1 func2
func1
func2
echo Done!
$ ksh autofunc.ksh
Testing autoload functions
KSH VER: Version AJMv 93u+m/1.1.0-alpha+42becab6 2021-08-13
typeset -fu func1
typeset -fu func2
func1 has executed.
func2 has worked.
Done!
$
```

**Comment by sowmya573** [comment]
On further investigation what we observed is, its not recreating in a simpler environment. So a simple recreate is not possible. 
We have an application where hundreds of scripts/functions are called one with in each other script. Only in such bigger environment, we are consistently seeing this issue. 

And also, other findings are:
the corruption is happening in bestreclaim() function. Once the autoload functions are created as nodes in the tree (shp->fdictpath), 

and when the flow goes to  this particular piece of code in xec.c:
     if(!np->nvalue.rp)                                                                                                        
                        {                                                                                                                         
                                np->nvalue.rp = new_of(struct Ufunction,shp->funload?sizeof(Dtlink_t):0);                                         
                                memset((void*)np->nvalue.rp,0,sizeof(struct Ufunction));                                                          
                        }      

new_of() calling bestreclaim() to utilize the freespace, is causing the issue. 
And particulary i see, in bestreclaim()
SELF(fp) = fp. --> this particular assignment is corrupting a node in the autload functions tree.

Any pointers or suggestions on how to proceed further?

Note: This is happening on ksh93u+ version(2012-08-01) on AIX

**Comment by hyenias** [comment]
I kindly suggest that you try using the current head for ksh93u+m as this is what this repo is about and then picking up your investigation work from there. There has been many change improvements since the 2012-08-01 ksh version. Please test with our current version of ksh.

**Comment by McDutchie** [comment]
Agreed with @hyenias.

@sowmya573, I appreciate how much effort went into your report and tracing, but a *lot* of development has happened here on the 93u+m fork. So many crashing bugs have been fixed since June 2020 that reporting crashes against 93u+ 2012-08-01 here is now pretty much a waste of time. This includes various fixes in the path search and function autoload code as well as a backported fix to dttree() itself (as part of cc4927529bb35557fdd772b8570d3855c68398d7).

Please download, build and test the beta release, or (even better) check out, build and test the current master branch, then see if your bug is still reproducible. If it is, please report it in a new issue. Thanks.
